### PR TITLE
Add Range node (for 412)

### PR DESCRIPTION
**Pull request by arporter**
2nd attempt at a PR to see if this resolves CodeCov's confusion.

### CODE REVIEW

**Comment by arporter** [comment]
It didn't help CodeCov which is still very confused about the diff it should be looking at. I've manually checked the coverage and I claim it is all good :-)
Ready for first review.

File: doc/developer_guide/psyir.rst [hunk 1]
```diff
@@ -168,6 +168,37 @@ Iteration construct
 .. autoclass:: psyclone.psyGen.Loop
     :members:
 
+Ranges
```
Comment (rupertford) [inline doc/developer_guide/psyir.rst]: Could you please update the user guide as appropriate.

File: doc/developer_guide/psyir.rst [hunk 2]
```diff
@@ -168,6 +168,37 @@ Iteration construct
...
+======
+
+The PSyIR has the `Range` node which represents a range of integer
+values with associated start, stop and step properties. e.g. the list
+of values [4, 6, 8, 10] would be represented by a `Range` with a start
+value of 4, a stop value of 10 and a step of 2 (all stored as `Literal`
+nodes).
```
Comment (rupertford) [inline doc/developer_guide/psyir.rst]: Say that this could be used for loop bounds and array slices but is not at the moment?

File: src/psyclone/psyGen.py [hunk 3]
```diff
@@ -6380,13 +6381,11 @@ def create(oper, lhs, rhs):
         '''Create a BinaryOperator instance given an operator and lhs and rhs
         child instances.
 
-        :param operator: the operator used in the operation.
-        :type operator: :py:class:`psyclone.psyGen.BinaryOperation.Operator`
-        :param lhs: the PSyIR node containing the left hand side of \
-            the assignment.
+        :param oper: the operator used in the operation.
```
Comment (rupertford) [inline src/psyclone/psyGen.py]: Thanks for fixing.

File: src/psyclone/psyir/nodes/ranges.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+
```
Comment (rupertford) [inline src/psyclone/psyir/nodes/ranges.py]: Your name and affiliation please.

File: src/psyclone/psyir/nodes/ranges.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+''' Module containing definition of the Range node. '''
```
Comment (rupertford) [inline src/psyclone/psyir/nodes/ranges.py]: the definition?

File: src/psyclone/psyir/nodes/ranges.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+from psyclone.psyGen import Node, Literal, InternalError
+from psyclone.psyir.symbols import DataType
+
+
+class Range(Node):
+    '''Represents an integer, index range with an increment. As such it
+    has three children representing the start, stop and step
+    expressions.
+
+    :param ast: the entry in the fparser2 parse tree representing the code \
```
Comment (rupertford) [inline src/psyclone/psyir/nodes/ranges.py]: I think Sergi has an issue with calling it an ast. I suggest using node instead of ast.

File: src/psyclone/psyir/nodes/ranges.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+                contained within this directive or None.
+    :type ast: :py:class:`fparser.two.Fortran2003.Base` or NoneType
+    :param parent: PSyIR node that is the parent of this Range or None.
+    :type parent: :py:class:`psyclone.psyGen.Node` or NoneType
```
Comment (rupertford) [inline src/psyclone/psyir/nodes/ranges.py]: You don't document annotations.

File: src/psyclone/psyir/nodes/ranges.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        Create an internally-consistent Range object. If no step
+        is provided then it defaults to an integer Literal with value 1.
+
+        :param start: the PSyIR for the start value.
+        :type start: :py:class:`psyclone.psyGen.Node`
+        :param stop: the PSyIR for the stop value.
+        :type stop: :py:class:`psyclone.psyGen.Node`
+        :param step: the PSyIR for the increment/step or None.
+        :type step: :py:class:`psyclone.psyGen.Node` or NoneType
+        :param parent: the parent node of this Range in the PSyIR.
```
Comment (rupertford) [inline src/psyclone/psyir/nodes/ranges.py]: There is no type for parent.

File: src/psyclone/psyir/nodes/ranges.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+            step.parent = erange
+        else:
+            # No step supplied so default to a value of 1
+            erange.step = Literal("1", DataType.INTEGER, parent=erange)
+        return erange
+
+    @staticmethod
+    def _check_valid_input(value, name):
+        '''
+        Perform checks that the supplied value is valid as a child of an
```
Comment (rupertford) [inline src/psyclone/psyir/nodes/ranges.py]: Remove 'an'

File: src/psyclone/tests/psyir/nodes/test_range.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
```
Comment (rupertford) [inline src/psyclone/tests/psyir/nodes/test_range.py]: Name in lights please.

File: src/psyclone/tests/psyir/nodes/test_range.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+
+''' pytest tests for the Range class. '''
+
+from __future__ import absolute_import
+import pytest
+from psyclone.psyir.symbols import DataType
+from psyclone.psyir.nodes.ranges import Range
+from psyclone.psyGen import InternalError, Literal, Reference
+
+
```
Comment (rupertford) [inline src/psyclone/tests/psyir/nodes/test_range.py]: Please add a test for __init__ i.e. parent, ast, annotations and _children default to expected values if not provided and that values are stored as expected for parent, ast and annotations when they are provided.

Comment (rupertford) [reply src/psyclone/tests/psyir/nodes/test_range.py]: Also add a test for create with a parent node and check that it defaults to None.

File: doc/developer_guide/psyir.rst [hunk 12]
```diff
@@ -168,6 +168,37 @@ Iteration construct
...
+
+The `Range` node must also provide support for array-slicing
+constructs where a user may wish to represent the entire range of
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/developer_guide/psyir.rst]: Updated.

File: doc/developer_guide/psyir.rst (hunk 1 above)
Comment (arporter) [reply doc/developer_guide/psyir.rst]: That's a difficult one as I'm not entirely clear about the distinction between user- and developer-guides when it comes to the PSyIR. I've simply mentioned it as a node type in the UG.

File: doc/developer_guide/psyir.rst (hunk 2 above)
Comment (arporter) [reply doc/developer_guide/psyir.rst]: Yes, done.

File: src/psyclone/psyir/nodes/ranges.py (hunk 4 above)
Comment (arporter) [reply src/psyclone/psyir/nodes/ranges.py]: Oh yes, done.

File: src/psyclone/psyir/nodes/ranges.py (hunk 5 above)
Comment (arporter) [reply src/psyclone/psyir/nodes/ranges.py]: Mmm, that's very definite.

File: src/psyclone/psyir/nodes/ranges.py (hunk 6 above)
Comment (arporter) [reply src/psyclone/psyir/nodes/ranges.py]: Oh yes, it's an old name. I've gone for `parse_node`.

File: src/psyclone/psyir/nodes/ranges.py (hunk 7 above)
Comment (arporter) [reply src/psyclone/psyir/nodes/ranges.py]: Oops.

File: src/psyclone/psyir/nodes/ranges.py (hunk 8 above)
Comment (arporter) [reply src/psyclone/psyir/nodes/ranges.py]: Oops again.

File: src/psyclone/psyir/nodes/ranges.py (hunk 9 above)
Comment (arporter) [reply src/psyclone/psyir/nodes/ranges.py]: Changed.

File: src/psyclone/tests/psyir/nodes/test_range.py (hunk 10 above)
Comment (arporter) [reply src/psyclone/tests/psyir/nodes/test_range.py]: If you insist... :-)

File: src/psyclone/tests/psyir/nodes/test_range.py (hunk 11 above)
Comment (arporter) [reply src/psyclone/tests/psyir/nodes/test_range.py]: Done.

**Comment by arporter** [comment]
CodeCov is messed up again but this is ready for 2nd review.

**Comment by arporter** [comment]
Note that this PR also includes a minor fix for the dag-generation code that is currently broken on master (because of the re-factoring of the various node classes).

File: doc/developer_guide/psyir.rst (hunk 1 above)
Comment (rupertford) [reply doc/developer_guide/psyir.rst]: Agreed, it is unclear. A mention if fine given the way we currently document.

File: src/psyclone/tests/psyir/nodes/test_range.py (hunk 11 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/nodes/test_range.py]: Great

File: src/psyclone/tests/psyir/nodes/test_range.py [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+    # When arguments are provided
+    parent = Node()
+    reader = FortranStringReader("program hello\nend program hello\n")
+    prog = parser(reader)
+    erange2 = Range(parse_node=prog, parent=parent)
+    assert erange2.parent is parent
+    assert erange2.ast is prog
+    with pytest.raises(InternalError) as err:
+        _ = Range(annotations=["was-where"])
+    # TODO correct US spelling of unrecognized!
```
Comment (rupertford) [inline src/psyclone/tests/psyir/nodes/test_range.py]: Please either do the todo or reference an issue.

File: src/psyclone/parse/utils.py [hunk 14]
```diff
@@ -89,12 +88,12 @@ def check_line_length(filename):
 
     :param str filename: The file containing the code.
 
-    :raises InternalError: if the specified file can not be opened or \
-    read.
+    :raises InternalError: if the specified file can not be opened or read.
```
Comment (rupertford) [inline src/psyclone/parse/utils.py]: Please update copyright date

File: src/psyclone/psyir/nodes/__init__.py [hunk 15]
```diff
@@ -48,6 +48,7 @@
 from psyclone.psyir.nodes.loop import Loop
 from psyclone.psyir.nodes.container import Container
 from psyclone.psyir.nodes.codeblock import CodeBlock
+from psyclone.psyir.nodes.ranges import Range
```
Comment (rupertford) [inline src/psyclone/psyir/nodes/__init__.py]: Please add your name as file modifier.

File: src/psyclone/psyir/nodes/__init__.py [hunk 16]
```diff
@@ -60,6 +61,7 @@
         'UnaryOperation',
         'BinaryOperation',
         'NaryOperation',
+        'Range',
```
Comment (rupertford) [inline src/psyclone/psyir/nodes/__init__.py]: My pylint complains about indentation (for all entries). If this is not the case for you then please ignore.

File: src/psyclone/tests/psyir/nodes/test_range.py (hunk 13 above)
Comment (arporter) [reply src/psyclone/tests/psyir/nodes/test_range.py]: Done it.

File: src/psyclone/parse/utils.py (hunk 14 above)
Comment (arporter) [reply src/psyclone/parse/utils.py]: Done.

File: src/psyclone/psyir/nodes/__init__.py (hunk 15 above)
Comment (arporter) [reply src/psyclone/psyir/nodes/__init__.py]: Done.

File: src/psyclone/psyir/nodes/__init__.py (hunk 16 above)
Comment (arporter) [reply src/psyclone/psyir/nodes/__init__.py]: My pylint is very happy!

**Comment by arporter** [comment]
Assuming Travis is happy, this is ready for 3rd review...

File: src/psyclone/psyir/nodes/__init__.py (hunk 16 above)
Comment (rupertford) [reply src/psyclone/psyir/nodes/__init__.py]: OK.
